lib/Doctrine/DBAL/Driver/PDOStatementImplementations.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+namespace Doctrine\DBAL\Driver;
+
+use Doctrine\DBAL\Driver\PDO\Exception;
+use PDOException;
+
+use function array_reduce;
+use function assert;
+use function func_get_args;
+use function is_array;
+
+use const PHP_VERSION_ID;
+
+if (PHP_VERSION_ID >= 80000) {
+    /**
+     * @internal
+     */
+    trait PDOStatementImplementations
+    {
+        /**
+         * {@inheritdoc}
+         *
+         * @deprecated Use one of the fetch- or iterate-related methods.
+         */
+        public function setFetchMode($mode, ...$args)
+        {
+            $mode = $this->convertFetchMode($mode);
+
+            $hasFetchModeArgs = array_reduce(
+                $args,
+                static function (bool $hasFetchModeArgs, mixed $value): bool {
+                    return $hasFetchModeArgs || $value !== null;
+                },
+                false
+            );
+            if (! $hasFetchModeArgs) {
+                $args = [];
+            }
+
+            try {
+                return parent::setFetchMode($mode, ...$args);
+            } catch (PDOException $exception) {
+                throw Exception::new($exception);
+            }
+        }
+
+        /**
+         * {@inheritdoc}
+         *
+         * @deprecated Use fetchAllNumeric(), fetchAllAssociative() or fetchFirstColumn() instead.
+         */
+        public function fetchAll($mode = null, ...$args)
+        {
+            if ($mode !== null) {
+                $mode = $this->convertFetchMode($mode);
+            }
+
+            $hasFetchModeArgs = array_reduce(
+                $args,
+                static function (bool $hasFetchModeArgs, mixed $value): bool {
+                    return $hasFetchModeArgs || $value !== null;
+                },
+                false
+            );
+            if (! $hasFetchModeArgs) {
+                $args = [];
+            }
+
+            try {
+                $data = parent::fetchAll($mode, ...$args);
+                assert(is_array($data));
+
+                return $data;
+            } catch (PDOException $exception) {
+                throw Exception::new($exception);
+            }
+        }
+    }
+} else {
+    /**
+     * @internal
+     */
+    trait PDOStatementImplementations
+    {
+        /**
+         * {@inheritdoc}
+         *
+         * @deprecated Use one of the fetch- or iterate-related methods.
+         */
+        public function setFetchMode($fetchMode, $arg2 = null, $arg3 = null)
```

Comment:
I was a bit irritated by the argument names, but looking at the method signature for `PDOStatement::setFetchMode()` I get why it is what it is 🙈

Reply:
The php 7 version of each trait is just C&P of the current implementation.